### PR TITLE
Address flaky tests

### LINKED ISSUE #379: [BUG] message handler doesn't verify if task handler has processed message
#### Describe the bug

When the daemon message handler receives a message that requires the task handler to do something (like start a process), the message is forwarded via a channel, as the task handler runs in a separate thread.

At no point is there any verification that the message has been handled by the task handler, before the message handler decides what response to give to the handled message.

This can lead to weird outcomes, where the message handler states that starting a task failed, but the task handler _later on_ starts the task anyway.

#### Steps to reproduce the bug

Check out my Flaky Test PR, #378, or apply the changes from the *Dev support: add test-log to help debug tests* commit in that PR to add `test-log` to the project + the `test(...)` attribute update for the `daemon::edit::test_edit_flow` test.

Run the test in a tight loop:

```sh
while RUST_LOG=debug cargo nextest run --workspace --hide-progress-bar --status-level=fail \
    daemon::edit::test_edit_flow
do :; done
```

At some point the test fails, but logs will show that the *Start* message failed because the task is locked for editing:

```
[2022-10-17T16:49:20Z DEBUG pueue_lib::network::protocol] Received message: Start(
        StartMessage {
            tasks: TaskIds(
                [
                    0,
                ],
            ),
            children: false,
        },
    )
[2022-10-17T16:49:20Z DEBUG pueue_lib::network::protocol] Sending message: Failure(
        "The command failed for tasks: 0",
    )
```

Between these to log entries, the message handler has sent the start message on to the task handler, and *immediately* produces a response from the current state.

In the case of a test failure, an `Edit` message is received by the daemon and also handled, putting the task into an unlocked, queued state.

However, the task handler has not yet obtained the mutex lock during all of this time. Even though the edit flow has been completed, the task handler finally can lock the state, finds the task in a queued state and so starts the task:

```
[2022-10-17T16:49:20Z INFO  pueue_daemon_lib::task_handler::spawn_task] Started task: ls -ahl
```

#### Expected behavior

The message handler should only produce a response once the task handler has processed the message too, otherwise it can produce responses that are not related to the actual state of the task handler.

#### Additional context

I can reproduce this issue on my Macbook Pro with M1 Pro CPU, I've not yet seen it when I run the test in a loop in a docker container running Linux. There is nothing in the implementation from preventing this bug from cropping up in Linux, however.

### PR DESCRIPTION
**Pull request by mjpieters**
Handle flaking tests better at the CI level, make it easier to find flaking tests and address some of those issues.

This PR extracts work done to mitigate flaky tests from #376. I'll rebase that PR on top of this one (so include the commits here into that PR). If this PR is merged before the other, the other can be rebased again trivially.

Fixes #379.

## Background

Some tests in the test suite flake, fail based on exact timings or because of issues with test helpers. These failures are not signs that the project code is broken and so should not block accepting new work.

This PR aims to update how tests are run on GitHub Actions so flaky tests can be detected and be re-run to verify that the code-under-test is not the source of the test failure.

## Changes included in the PR

1. Switch to [cargo nextest](https://nexte.st/) to run the test suite. Not only does nextest run the test suite faster, it also can re-run tests that failed to verify that the failure wasn't a flake.

2. Adds a [test reporter action](https://github.com/marketplace/actions/publish-test-results), which surfaces test results in PRs and GitHub commit contexts. This makes it possible to see that a test flaked, as flaky tests, ideally, are fixed so they can't flake.

  **Note**: This part requires an additional GitHub workflow to actually publish the test report. This workflow will *not run* while still confined in this PR, because GitHub will not yet be able to find the workflow in the default
  (main) branch until it is merged into *that branch*. I've created a [demonstration PR](https://github.com/mjpieters/pueue/pull/2) in my own fork that shows what these reports look like.

3. Update dev-dependencies to make it easier to debug flaking tests, by making
   it trivially easy to enable logging for specific tests. I have found the
   existing logging already present in the pueue codebase to be _excellent_,
   and now only a `RUST_LOG=debug` environment variable away from becoming
   visible when a test fails.

4. And finally, individual commits to fix specific flaking tests I found I was
   able to fix.

## How to find and debug flaking tests

Run the test suite in a continuous loop; preferably with cargo nextest as that's faster; I've tuned the command-line switches to minimise output for runs without test failures:

```sh
while cargo nextest run --workspace --hide-progress-bar --status-level=fail
do :; done
```

You can do the same with the default `cargo test` command:

```sh
while cargo test --quiet
do :; done
```

Once a flaking test has been identified, add a test filter to focus on that specific test, add `RUST_LOG=debug` to the command-line to ensure you get plenty of log output when a failure happens, and if the test in question hasn't already been updated to [use the test-log macro](https://crates.io/crates/test-log), do so before trying to reproduce the issue:

```sh
while RUST_LOG=debug cargo nextest run --workspace --hide-progress-bar --status-level=fail \
    daemon::edit::test_edit_flow
do :; done
```

The cargo test equivalent is:

```sh
while RUST_LOG=debug cargo test --workspace --quiet \
    --test daemon_tests daemon::edit::test_edit_flow
do :; done
```

When you next hit a failure, you should have a plethora of information as to what pueue was doing up to that point.

## Checklist

- [x] I picked the correct source and target branch.
- [x] I included a new entry to the `CHANGELOG.md`.
- [x] I checked `cargo clippy` and `cargo fmt`. The CI will fail otherwise anyway.
- [ ] (If applicable) I added tests for this feature or adjusted existing tests.
- [ ] (If applicable) I checked if anything in the wiki needs to be changed.

### CODE REVIEW

**Comment by mjpieters** [comment]
The only check failure here, raised by codecov, will most likely go away if #382 were to be merged (and I rebase this PR).

File: .github/workflows/test.yml
Comment (Nukesor) [inline .github/workflows/test.yml:125]: If i understand this correctly, this uploads the full webhook event, that would be send if there were any webhooks specified?
This is then unpacked in the test-report action and used in the publish-unit-test-result-action call.

It would be great to have a few lines of docs on why this is desirable.

File: tests/daemon/edit.rs
Comment (Nukesor) [inline tests/daemon/edit.rs:30]: Would you recommend to leave this in here?
If so, should we consider adding this to all tests?

File: Cargo.toml
Comment (Nukesor) [inline Cargo.toml:95]: Maybe we should start a new document or a wiki page for development practices.
Your tutorial on how to catch and debug flaky tests would also be good to have in there :)

**Review by Nukesor — CHANGES_REQUESTED** [review]
Nice.

Overall this looks good to me :) Thanks for tackling those flaky tests.
I didn't expect to have that many flaky tests left in this codebase :sweat_smile: 

Just gave nextest a spin and it looks great! Especially the flakiness detection is a godsend. I'll probably introduduce this i other projects as well!

File: .github/workflows/test-report.yml
Comment (mjpieters) [reply .github/workflows/test-report.yml:6]: This allows the report runner to create a report even if the tests were run against a forked repo (e.g. any PR from someone other than you).

See the [documentation for the report action](https://github.com/marketplace/actions/publish-test-results#support-fork-repositories-and-dependabot-branches); the separate workflow is run whenever the test workflow completes, but it is run within the pueue repo, always. The downside is that it must be present in the default branch for GitHub to find it.

File: .github/workflows/test-report.yml
Comment (mjpieters) [reply .github/workflows/test-report.yml:36]: Sorry, it's a straight copy-pasta from the [Publish Test Results documentation](https://github.com/marketplace/actions/publish-test-results#support-fork-repositories-and-dependabot-branches).

Happy to add a comment or two anyway :-)

File: .github/workflows/test.yml
Comment (mjpieters) [reply .github/workflows/test.yml:122]: Missed this one, I spotted the other when I copied the sample workflow setup 🤦

File: tests/daemon/edit.rs
Comment (mjpieters) [reply tests/daemon/edit.rs:30]: It shouldn't make much difference if left in place. It'll be helpful for future debugging issues.

I recommend adding it to tests where you feel the need to see log output, on a case by case basis. There is little point in adding it to all the tests just for the sake of it.

File: Cargo.toml
Comment (mjpieters) [reply Cargo.toml:95]: Sounds like a plan! Perhaps add a DEVELOPMENT.md pointing to such a page?

File: .github/workflows/test.yml
Comment (mjpieters) [reply .github/workflows/test.yml:125]: Again, it's a requirement set by the publishing action, as it uses the event file as a basis. The separate workflow handling the reports means it would otherwise have the wrong context.

File: .github/workflows/test-report.yml
Comment (mjpieters) [reply .github/workflows/test-report.yml:30]: I suspect that the downloaded would have the wrong context, as the artefacts were uploaded by a different workflow.

The bash script takes the events file info instead. 

This is prescribed by the report action documentation.

File: Cargo.toml
Comment (Nukesor) [reply Cargo.toml:95]: Sounds good to me!

File: .github/workflows/test.yml
Comment (mjpieters) [reply .github/workflows/test.yml:112]: Looks like I misread when `success()` is `true`, I thought it would be false when the workflow was cancelled. Re-reading the docs it is entirely redundant here.

File: Cargo.toml
Comment (mjpieters) [reply Cargo.toml:95]: I've made a start here: https://github.com/Nukesor/pueue/wiki/Development

Shall I leave it up to you to fill in a DEVELOPMENT.md file?

Comment (Nukesor) [reply Cargo.toml:95]: I'll extend the wiki entry and transfer it to the DEVELOPMENT.md lateron :)

**Review by Nukesor — APPROVED** [review]
Nice. I thinks we are good to merge this :+1: 

There seem to be a few merge conflicts with your other branch. I'll merge once this is done :)

**Comment by mjpieters** [comment]
> There seem to have a few merge conflicts with your other branch though. I'll merge once this is done :)

Rebase done!

**Pull request merged**
